subprojects/base-services/src/main/java/org/gradle/internal/classpath/TransformedClassPath.java

```diff
@@ -65,53 +85,90 @@ public List<File> getAsTransformedFiles() {
         return originals;
     }
 
+    /**
+     * {@inheritDoc}
+     * <p>
+     * This method returns the list of original JAR/class directory URLs.
+     */
     @Override
     public List<URL> getAsURLs() {
         return originalClassPath.getAsURLs();
     }
 
+    /**
+     * {@inheritDoc}
+     * <p>
+     * This method returns the array of original JAR/class directory URLs.
+     */
     @Override
     public URL[] getAsURLArray() {
         return originalClassPath.getAsURLArray();
     }
 
-    ClassPath prepend(DefaultClassPath classPath) {
-        return new TransformedClassPath(classPath.plus(originalClassPath), transforms);
+    /**
+     * Prepends the given classPath to this classpath.
+     * No transformations are applied to the prepended class path.
+     * <p>
+     * This is a helper to implement DefaultClassPath.plus(TransformedClassPath) efficiently.
+     *
+     * @param classPath the classPath to prepend to this classpath
+     * @return the new transformed classpath
+     */
+    TransformedClassPath prepend(DefaultClassPath classPath) {
+        // If some entries from this classpath are also in the prepended classpath, then the transform should be undone in the result.
```

Comment:
❓ I don't get _then the transform should be undone in the result_, could you rephrase or expand?

Reply:
I've tried to give a more detailed explanation in the class javadoc. It should also reduce the confusion in the test case https://github.com/gradle/gradle/pull/23351#discussion_r1086626689, as these are actually closely related.